pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -148,8 +152,32 @@ public synchronized void registerSendOp(CompletableFuture<MessageId> sendFuture)
         }
     }
 
+    public synchronized <T> void registerInternal(CompletableFuture<T> completableFuture) {
+        //There have been an exception, it means this transaction should not be commit,
+        //so there is no need to record the execution of other operations
+        if (!executedFuture.isCompletedExceptionally()) {
+            opsExecutingInTxn.incrementAndGet();
+            executedFuture = new CompletableFuture<>();
+            completableFuture.thenRun(() -> {
+                synchronized (this) {
+                    opsExecutingInTxn.decrementAndGet();
+                    // This is the last operation so far.
+                    if (opsExecutingInTxn.get() == 0) {
+                        executedFuture.complete(null);
+                    }
+                }
+            }).exceptionally(e -> {
+                //Complete this future exceptionally and there is no need to executed this method again.
+                synchronized (this) {
```

Comment:
Why need `synchronized` here.

Reply:
When a thread passes the if judgment and prepares  to new a future, if an exception completely here, it will be ignored.

Reply:
So we hope `executedFuture = new CompletableFuture<>();`, `executedFuture.complete(null);` and `executedFuture.completeExceptionally` should be exclusive.